ci: Detect breaking changes in packages other than Uno.[UI|WinUI]

Youssef1313 opened the pull request:
GitHub Issue (If applicable): closes #11548, closes https://github.com/unoplatform/uno/issues/10928

<!-- Link to relevant GitHub issue if applicable. All PRs should be associated with an issue (GitHub issue or internal), unless the change is documentation related. -->

## PR Type

What kind of change does this PR introduce?

- Build or CI related changes

## What is the current behavior?

Breaking changes are not detected for packages other than Uno.UI and Uno.WinUI.

## What is the new behavior?

They are detected now.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
     Please note that breaking changes are likely to be rejected -->

## Other information

<!-- Please provide any additional information if necessary -->

Internal Issue (If applicable):
<!-- Link to relevant internal issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

Youssef1313 commented on src/Uno.UI.Runtime.Skia.Wpf/Controls/WpfTextViewTextBox.cs:
```diff
@@ -3,12 +3,12 @@
 namespace Uno.UI.Runtime.Skia.Wpf.Controls;
 
-internal class WpfTextViewTextBox : TextBox
+internal class WpfTextViewTextBox2 : TextBox
```
Oh this **was** public in past. I will find another public type to test the change on.

Youssef1313 commented:
It's working!

```
  Comparing C:\Users\AzDevOps\AppData\Local\Temp\8b6e03e7-2268-4bf8-9adc-7b1e437b3229\lib\net47\Uno.UI.Runtime.Skia.Wpf.dll and C:\Users\AzDevOps\AppData\Local\Temp\6fe28c55-9ad6-4394-a5f3-0effcbd437c5\lib\net47\Uno.UI.Runtime.Skia.Wpf.dll
##[error]EXEC(0,0): Error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set.
EXEC : error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set. [C:\a\1\s\Build\Uno.UI.Build.csproj]
  Comparing C:\Users\AzDevOps\AppData\Local\Temp\8b6e03e7-2268-4bf8-9adc-7b1e437b3229\lib\net7.0-windows7.0\Uno.UI.Runtime.Skia.Wpf.dll and C:\Users\AzDevOps\AppData\Local\Temp\6fe28c55-9ad6-4394-a5f3-0effcbd437c5\lib\net7.0-windows7.0\Uno.UI.Runtime.Skia.Wpf.dll
##[error]EXEC(0,0): Error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set.
EXEC : error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set. [C:\a\1\s\Build\Uno.UI.Build.csproj]
  Comparing C:\Users\AzDevOps\AppData\Local\Temp\8b6e03e7-2268-4bf8-9adc-7b1e437b3229\lib\netcoreapp3.1\Uno.UI.Runtime.Skia.Wpf.dll and C:\Users\AzDevOps\AppData\Local\Temp\6fe28c55-9ad6-4394-a5f3-0effcbd437c5\lib\netcoreapp3.1\Uno.UI.Runtime.Skia.Wpf.dll
##[error]EXEC(0,0): Error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set.
EXEC : error : Removed type Uno.UI.Runtime.Skia.Wpf.WpfApplicationExtension not found in ignore set. [C:\a\1\s\Build\Uno.UI.Build.csproj]
  Done comparing, with differences.
##[error]EXEC(0,0): Error : Build failed with unexpected differences. Modifications to the public API introduce binary breaking changes and should be avoided.
EXEC : error : Build failed with unexpected differences. Modifications to the public API introduce binary breaking changes and should be avoided. [C:\a\1\s\Build\Uno.UI.Build.csproj]
  If these modifications were expected and intended, see https://github.com/nventive/Uno.PackageDiff#how-to-provide-an-ignore-set on how to ignore them.
##[error]Build\Uno.UI.Build.csproj(499,3): Error MSB3073: The command "C:\a\1\s\Build\\tools\generatepkgdiff.exe --base=Uno.UI.Runtime.Skia.Wpf --other=..\..\..\NugetPackages-Artifacts-skia-UWP\vslatest-skia\Uno.UI.Runtime.Skia.Wpf.4.9.0-PullRequest-11845-74211-11845.nupkg --diffignore=PackageDiffIgnore.xml --outfile=C:\a\1\a\ApiDiff.Uno.UI.Runtime.Skia.Wpf.4.9.0-PullRequest-11845-74211-11845.md" exited with code 1.
C:\a\1\s\Build\Uno.UI.Build.csproj(499,3): error MSB3073: The command "C:\a\1\s\Build\\tools\generatepkgdiff.exe --base=Uno.UI.Runtime.Skia.Wpf --other=..\..\..\NugetPackages-Artifacts-skia-UWP\vslatest-skia\Uno.UI.Runtime.Skia.Wpf.4.9.0-PullRequest-11845-74211-11845.nupkg --diffignore=PackageDiffIgnore.xml --outfile=C:\a\1\a\ApiDiff.Uno.UI.Runtime.Skia.Wpf.4.9.0-PullRequest-11845-74211-11845.md" exited with code 1.
##[error]Process 'msbuild.exe' exited with code '1'.
```

Youssef1313 commented on src/Uno.UI.Runtime.Skia.Wpf/WpfApplicationExtension.cs:
```diff
@@ -7,15 +7,15 @@
 namespace Uno.UI.Runtime.Skia.Wpf
 {
-	public class WpfApplicationExtension : IApplicationExtension, IDisposable
+	public class WpfApplicationExtension2 : IApplicationExtension, IDisposable
```
```suggestion
	public class WpfApplicationExtension: IApplicationExtension, IDisposable
```

Youssef1313 commented on src/Uno.UI.Runtime.Skia.Wpf/WpfApplicationExtension.cs:
```diff
@@ -7,15 +7,15 @@
 		private bool _disposedValue;
 
-		public WpfApplicationExtension(Application owner)
+		public WpfApplicationExtension2(Application owner)
```
```suggestion
		public WpfApplicationExtension(Application owner)
```

Youssef1313 commented on src/Uno.UI.Runtime.Skia.Wpf/WpfApplicationExtension.cs:
```diff
@@ -54,7 +54,7 @@ protected virtual void Dispose(bool disposing)
 		}
 
-		~WpfApplicationExtension()
+		~WpfApplicationExtension2()
```
```suggestion
		~WpfApplicationExtension()
```

Youssef1313 commented on src/Uno.UI.Runtime.Skia.Wpf/WpfHost.cs:
```diff
@@ -84,7 +84,7 @@ internal static void RegisterExtensions()
 			ApiExtensibility.Register(typeof(Uno.ApplicationModel.Core.ICoreApplicationExtension), o => new CoreApplicationExtension(o));
 			ApiExtensibility.Register(typeof(Windows.UI.Core.ICoreWindowExtension), o => new WpfCoreWindowExtension(o));
-			ApiExtensibility.Register<Windows.UI.Xaml.Application>(typeof(IApplicationExtension), o => new WpfApplicationExtension(o));
+			ApiExtensibility.Register<Windows.UI.Xaml.Application>(typeof(IApplicationExtension), o => new WpfApplicationExtension2(o));
```
```suggestion
			ApiExtensibility.Register<Windows.UI.Xaml.Application>(typeof(IApplicationExtension), o => new WpfApplicationExtension(o));
```

jeromelaban commented on build/Uno.UI.Build.csproj:
```diff
@@ -482,7 +482,28 @@
 
 		<Exec Command="dotnet tool install --tool-path $(MSBuildThisFileDirectory)\tools Uno.PackageDiff --version 1.1.0-dev.14" IgnoreExitCode="true" />
-		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix) --other=$(PackageNamePrefix).$(GITVERSION_SemVer).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.$(GITVERSION_SemVer).md" />
+		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix) --other=$(PackageNamePrefix).$(GITVERSION_SemVer).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.$(PackageNamePrefix).$(GITVERSION_SemVer).md" />
+		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix).Foldable --other=..\..\..\NugetPackages-Artifacts-generic-UWP\vslatest-generic\$(PackageNamePrefix).Foldable.$(GITVERSION_SemVer).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.$(PackageNamePrefix).Foldable.$(GITVERSION_SemVer).md" />
+		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix).Lottie --other=$(PackageNamePrefix).Lottie.$(GITVERSION_SemVer).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.$(PackageNamePrefix).Lottie.$(GITVERSION_SemVer).md" />
+		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix).Maps --other=..\..\..\NugetPackages-Artifacts-generic-UWP\vslatest-generic\$(PackageNamePrefix).Maps.$(GITVERSION_SemVer).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.$(PackageNamePrefix).Maps.$(GITVERSION_SemVer).md" />
+		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix).MSAL --other=$(PackageNamePrefix).MSAL.$(GITVERSION_SemVer).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.$(PackageNamePrefix).MSAL.$(GITVERSION_SemVer).md" />
```
We should be able to use something like that:

```suggestion
        <ItemGroup>
           <_diffPackage Include="$(PackageNamePrefix).$(GITVERSION_SemVer)" />
            <_diffPackage Include="$(PackageNamePrefix).Foldable.$(GITVERSION_SemVer)" />
        </ItemGroup>

		<Exec Command="$(MSBuildThisFileDirectory)\tools\generatepkgdiff.exe --base=$(PackageNamePrefix) --other=%(_diffPackage .Identity).nupkg --diffignore=PackageDiffIgnore.xml --outfile=$(OutputDir)\ApiDiff.%(_diffPackage .Identity).md" />
```

jeromelaban reviewed (approved):
🚀

Youssef1313 commented on build/Uno.UI.Build.csproj:
```diff
@@ -481,8 +481,33 @@
+			<_diffPackage Include="$(PackageNamePrefix).Maps" Other="..\..\..\NugetPackages-Artifacts-generic-UWP\vslatest-generic\$(PackageNamePrefix).Maps.$(GITVERSION_SemVer).nupkg" />
+			<_diffPackage Include="$(PackageNamePrefix).MSAL" Other="$(PackageNamePrefix).MSAL.$(GITVERSION_SemVer).nupkg" />
+			<_diffPackage Include="$(PackageNamePrefix).Svg" Other="$(PackageNamePrefix).Svg.$(GITVERSION_SemVer).nupkg" />
+			<_diffPackage Include="$(PackageNamePrefix).RemoteControl" Other="$(PackageNamePrefix).RemoteControl.$(GITVERSION_SemVer).nupkg" />
```
I assume I should remove RemoteControl as its APIs are not directly used by customers and we should freely change it? @jeromelaban Can you confirm please?

Youssef1313 replied:
Removed for now. We can add if necessary later.

Youssef1313 commented:
This is failing with:

```
EXEC(0,0): Error : Removed method System.Uri Microsoft.Toolkit.Uwp.UI.Lottie.LottieVisualSourceBase.get_UriSource() not found in ignore set.
EXEC(0,0): Error : Removed method System.Void Microsoft.Toolkit.Uwp.UI.Lottie.LottieVisualSourceBase.set_UriSource(System.Uri value) not found in ignore set.
```

I don't really understand, UriSource doesn't seem to be removed https://github.com/unoplatform/uno/blob/bdd49d9b5ec46c205c67b4903bfa9e5d99702627/src/AddIns/Uno.UI.Lottie/LottieVisualSourceBase.cs#L42-L46

@jeromelaban @Robert-Louis Any idea?

jeromelaban commented:
> @jeromelaban @Robert-Louis Any idea?

This is likely caused by a change in the implementation. `UriSource` is now a member of an interface (where it previously was not), and it's likely changing its actual definition.

Robert-Louis commented:
> This is likely caused by a change in the implementation. `UriSource` is now a member of an interface (where it previously was not), and it's likely changing its actual definition.

@jeromelaban how should we fix it ? Should I find a way to not adding it to the interface ? 
I not sure how the CI work with this.

Youssef1313 commented:
@jeromelaban That was my first guess, but looking into IL for a "simplified" case, I see that the call in both these cases produce the same IL

![image](https://user-images.githubusercontent.com/31348972/234021524-3ba2b432-b30e-4978-9b98-6ea9f18b45ed.png)

Maybe I'm missing something?
